tvb_contrib/tvb/contrib/cosimulation/cosim_monitors.py

```diff
@@ -0,0 +1,185 @@
+    It collects:
+
+        - all state variables and modes from class :Model:
+        - all nodes of a region or surface based
+        - all the integration time steps
+
+    """
+    _ui_name = "Cosimulation Raw recording"
+
+    def sample(self, start_step, n_steps, cosim_history, history):
+        "Return all the states of the partial (up to synchronization time) cosimulation history"
+        return self.get_sample(start_step, n_steps, cosim_history, cosim=True)
+
+
+class RawVoiCosim(RawVoi, CosimMonitor):
+    """
+    A monitor that records the output raw data of selected variables
+    from the partial (up to synchronization time) history of TVB simulation.
+    It collects:
+
+        - voi state variables and all modes from class :Model:
+        - all nodes of a region or surface based
+        - all the integration time steps
+
+    """
+    _ui_name = "Cosimulation RawVoi recording"
+
+    def sample(self, start_step, n_steps, cosim_history, history):
+        "Return all the states of the partial (up to synchronization time) cosimulation history"
+        return self.get_sample(start_step, n_steps, cosim_history, cosim=True)
+
+
+class RawDelayed(Raw, CosimMonitor):
+    """
+    A monitor that records the output raw data of all coupling variables
+    from the full history of a TVB simulation.
+    It collects:
+
+        - all coupling state variables and modes from class :Model:
+        - all nodes of a region or surface based
+        - all the integration time steps
+
+    """
+
+    _ui_name = "Cosimulation Raw Delayed recording"
+
+    def sample(self, start_step, n_steps, cosim_history, history):
+        "Return all the states of the delayed (by synchronization time) TVB history"
+        return self.get_sample(start_step, n_steps, history, cosim=False)
+
+
+class RawVoiDelayed(RawVoi,CosimMonitor):
+    """
+    A monitor that records the output raw data of selected coupling variables
+    from the full history of a TVB simulation.
+    It collects:
+
+        - selected coupling state variables and all modes from class :Model:
+        - all nodes of a region or surface based
+        - all the integration time steps
+
+    """
+
+    _ui_name = "Cosimulation RawVoi Delayed recording"
+
+    def sample(self, start_step, n_steps, cosim_history, history):
+        "Return selected states of the delayed (by synchronization time) TVB history"
+        return self.get_sample(start_step, n_steps, history, cosim=False)
+
+
+class CosimCoupling(AfferentCoupling, CosimMonitorFromCoupling):
+    """
+    A monitor that records the future coupling of selected variables:
+    It collects:
+
+        - selected coupling values and all modes from class :Model:
+        - all nodes of a region or surface based
+        - all the integration time steps
+
+    !!!WARNING don't use this monitor for a time smaller than the synchronization variable!!!
```

Comment:
as before please add runtime check, not docstring warning

Reply:
see above